mitmproxy/websocket.py

```diff
@@ -0,0 +1,83 @@
+import time
+
+from typing import List
+
+from mitmproxy import flow
+from mitmproxy.http import HTTPFlow
+from mitmproxy.net import websockets
+from mitmproxy.utils import strutils
+from mitmproxy.types import serializable
+
+
+class WebSocketMessage(serializable.Serializable):
+
+    def __init__(self, flow, from_client, content, timestamp=None):
+        self.flow = flow
+        self.content = content
+        self.from_client = from_client
+        self.timestamp = timestamp or time.time()
+
+    @classmethod
+    def from_state(cls, state):
+        return cls(*state)
+
+    def get_state(self):
+        return self.from_client, self.content, self.timestamp
+
+    def set_state(self, state):
+        self.from_client = state.pop("from_client")
+        self.content = state.pop("content")
+        self.timestamp = state.pop("timestamp")
+
+    @property
+    def info(self):
+        return "{client} {direction} WebSocket {type} message {direction} {server}{endpoint}".format(
+            type=self.type,
+            client=repr(self.flow.client_conn.address),
+            server=repr(self.flow.server_conn.address),
+            direction="->" if self.from_client else "<-",
+            endpoint=self.flow.handshake_flow.request.path,
+        )
+
+
+class WebSocketBinaryMessage(WebSocketMessage):
+
+    type = 'binary'
+
+    def __repr__(self):
+        return "binary message: {}".format(strutils.bytes_to_escaped_str(self.content))
+
+class WebSocketTextMessage(WebSocketMessage):
+
+    type = 'text'
+
+    def __repr__(self):
+        return "text message: {}".format(repr(self.content))
+
+
+class WebSocketFlow(flow.Flow):
```

Comment:
Thinking about it for a couple of nights:

I think if we have the handshake flow as the main thing, and the websocket messages **only** as additional members, it might be too hidden to work with. For me the fully established websocket is the main entity - which should contain the handshake information.

If one would want to manipulate the handshake and force other options, there is the `websocket_handshake` hook to do just that. After establishing the websocket connection, we follow the same procedure as RawTCP already does - this gives us a very nice and clean symmetry (in terms of how the protocols work).

Promoting the handshake flow to become a WebSocketFlow (after a successful handshake only) sounds nice. However, this entails quite a lot of other refactoring, which I don't want to touch right now in this PR.

This will be especially useful in the following cases:
  * HTTP/1.1 upgrade to WebSocket
  * HTTP/1.1 upgrade to HTTP/2
  * HTTP/1.1 change to RawTCP (e.g. byte-by-byte streaming, or to ignore individual paths, large file up-/downloads)

I haven't made up my mind yet if this will be of any real benefit - but it sounds nice 😄 

The mitmproxy core is in a state of flux at the moment - so I don't want to open up another major architectural change (at least not in this PR by myself). IMHO we should try to get this PR merged (in some form or another thats closer to what we already have), and then iterate to get a new flow architecture that allows us to promote / decorate / whatever.